### ISSUE TITLE
Cannot install `noarch: True` packages with `conda 4.3.8`

### ISSUE THREAD

**Issue by mingwandroid** [conda/conda#4428]
Attempting to install `noarch: True` packages, for example:

`conda install -c rdonnellyr font-ttf-inconsolata`

.. fails with:

```
Fetching package metadata ...........
Solving package specifications: .

Package plan for installation in environment /Users/rdonnelly/mc-x64-3.5:

The following NEW packages will be INSTALLED:

    font-ttf-inconsolata: 2.000-0 rdonnellyr

Proceed ([y]/n)? y

font-ttf-incon 100% |##############################################################################################################################################################################################################################| Time: 0:00:00 219.96 kB/s


CondaUpgradeError: Conda upgrade error: The current version of conda is too old to install this package.
Please update conda.
```

`conda --version`

.. gives:

`conda 4.3.8`

adding `-v -v -v` gives the following callstack:

```
    Traceback (most recent call last):
      File "/Users/rdonnelly/mc-x64-3.5/lib/python3.5/site-packages/conda/exceptions.py", line 617, in conda_exception_handler
        return_value = func(*args, **kwargs)
      File "/Users/rdonnelly/mc-x64-3.5/lib/python3.5/site-packages/conda/cli/main.py", line 137, in _main
        exit_code = args.func(args, p)
      File "/Users/rdonnelly/mc-x64-3.5/lib/python3.5/site-packages/conda/cli/main_install.py", line 80, in execute
        install(args, parser, 'install')
      File "/Users/rdonnelly/mc-x64-3.5/lib/python3.5/site-packages/conda/cli/install.py", line 347, in install
        execute_actions(actions, index, verbose=not context.quiet)
      File "/Users/rdonnelly/mc-x64-3.5/lib/python3.5/site-packages/conda/plan.py", line 837, in execute_actions
        execute_instructions(plan, index, verbose)
      File "/Users/rdonnelly/mc-x64-3.5/lib/python3.5/site-packages/conda/instructions.py", line 258, in execute_instructions
        cmd(state, arg)
      File "/Users/rdonnelly/mc-x64-3.5/lib/python3.5/site-packages/conda/instructions.py", line 119, in UNLINKLINKTRANSACTION_CMD
        txn.execute()
      File "/Users/rdonnelly/mc-x64-3.5/lib/python3.5/site-packages/conda/core/link.py", line 255, in execute
        self.verify()
      File "/Users/rdonnelly/mc-x64-3.5/lib/python3.5/site-packages/conda/core/link.py", line 238, in verify
        self.prepare()
      File "/Users/rdonnelly/mc-x64-3.5/lib/python3.5/site-packages/conda/core/link.py", line 169, in prepare
        concatv(unlink_actions, link_actions))
      File "/Users/rdonnelly/mc-x64-3.5/lib/python3.5/site-packages/conda/core/link.py", line 168, in <genexpr>
        self.all_actions = tuple(per_pkg_actions for per_pkg_actions in
      File "/Users/rdonnelly/mc-x64-3.5/lib/python3.5/site-packages/conda/core/link.py", line 165, in <genexpr>
        for pkg_info, lt in zip(self.packages_info_to_link, link_types)
      File "/Users/rdonnelly/mc-x64-3.5/lib/python3.5/site-packages/conda/core/link.py", line 387, in make_link_actions
        file_link_actions = LinkPathAction.create_file_link_actions(*required_quad)
      File "/Users/rdonnelly/mc-x64-3.5/lib/python3.5/site-packages/conda/core/path_actions.py", line 179, in create_file_link_actions
        return tuple(make_file_link_action(spi) for spi in package_info.paths_data.paths)
      File "/Users/rdonnelly/mc-x64-3.5/lib/python3.5/site-packages/conda/core/path_actions.py", line 179, in <genexpr>
        return tuple(make_file_link_action(spi) for spi in package_info.paths_data.paths)
      File "/Users/rdonnelly/mc-x64-3.5/lib/python3.5/site-packages/conda/core/path_actions.py", line 150, in make_file_link_action
        Please update conda."""))
    conda.exceptions.CondaUpgradeError: Conda upgrade error: The current version of conda is too old to install this package.
    Please update conda.
```

.. adding a debug print to that exception (` (noarch is {})""".format(noarch)`) gives:

```
conda.exceptions.CondaUpgradeError: Conda upgrade error: The current version of conda is too old to install this package.
Please update conda. (noarch is Noarch(_Noarch__initd=True, type='True'))
```

**Comment by kalefranz** [comment]
I think the code as it is right now understands and respects `noarch: generic` if you don't mind updating the metadata in the package.  We can also think through respecting `noarch: True` if it's important to you.

**Comment by kalefranz** [comment]
I also don't think we have an integration test yet for `noarch: generic`.  I'll add one.

**Comment by mingwandroid** [comment]
I. implemented this originally and have released packages like this therefore there is no option but for conda to respect noarch: True

**Comment by kalefranz** [comment]
fixed with #4431

**Comment by gabm** [comment]
My users experience the same problem: Under Windows they can't install packages that were built with `noarch: True`. Under Linux however it seems to work.

Writing here just to make sure the policy: Will conda continue to support `noarch: True` in future or do we have to re-write all recipes? I don't like condas recent policy to break things that already worked..

Also the userguide should be updated: https://conda.io/docs/building/meta-yaml.html

**Comment by mingwandroid** [comment]
You will not have to rewrite your recipes, no.

@kalefranz, are you going to release a new version with this fix? It would be bad if Anaconda 4.3 contained this bug.

**Comment by mingwandroid** [comment]
Reopening until this fix makes it into a released version.

**Comment by gabm** [comment]
Aaah I see, ok: So with this PR now we have:

```
noarch: python
noarch: generic
noarch: True      # which means the same as generic
```

Is it safe to assume that new packages should use *generic* for the future?

**Comment by mingwandroid** [comment]
You are free to use either.

**Comment by Korijn** [comment]
@gabm would be very useful to see that little overview in the docs somewhere...

**Comment by gabm** [comment]
@Korijn I don't know where to contribute that.. but it would be useful indeed

@kalefranz @mingwandroid It seems to be fixed with 4.3.9...

**Comment by Korijn** [comment]
Neither do I!

**Comment by mingwandroid** [comment]
@kalefranz has a preference for `generic`. To contribute to conda's docs, please submit a PR to: https://github.com/conda/conda-docs

**Comment by gabm** [comment]
So... `generic` is the official way to go? I'll provide a PR.

Meanwhile I re-checked conda 4.3.9. Now it installs fine - but no files are installed on windows... I'll check on linux as well...

**Comment by kalefranz** [comment]
Fixed in #4431 with 4.3.9 released on Jan 31.

@gabm Python files on windows get installed into `$PREFIX/Lib/site-packages`, not `$PREFIX/Library`.  Please file a new issue if you think there's still a problem and I'll get right on it.
